src/basic/efivars.c

```diff
@@ -222,4 +222,29 @@ int efi_set_variable_string(sd_id128_t vendor, const char *name, const char *v)
         return efi_set_variable(vendor, name, u16, (char16_strlen(u16) + 1) * sizeof(char16_t));
 }
 
+int efi_options_variable(char **line) {
+        const char *e;
+        int r;
+
+        assert(line);
+
+        /* For testing purposes it is sometimes useful to be able to override this */
+        e = secure_getenv("SYSTEMD_EFI_OPTIONS");
+        if (e) {
+                char *m;
+
+                m = strdup(e);
+                if (!m)
+                        return -ENOMEM;
+
+                *line = m;
+                return 0;
+        }
+
+        r = efi_get_variable_string(EFI_VENDOR_LOADER, "SystemdOptions", line);
```

Comment:
hmm, shouldn't this maybe something the loader appends, not userspace?

i mean, in particular as you put this in the loader UEFI UUID namespace?

The capitalization of "SystemdOptions" is a bit weird...

Shouldn't this be KernelCommandLineAppend or so, and then simply be appended by the boot loader or the EFI stub? I mean, otherwise people would try to add "audit=0 selinux=0" and then be very annoyed and surprised if it doesn't work.

Also, I am pretty sure this should also come in a oneshot flavour, so that you can append "fsck.mode=force" for the next boot only.

Reply:
> hmm, shouldn't this maybe something the loader appends, not userspace?

I wanted this to work independently of the loader. In particular, I'm thinking about providing an alternative way to say 'systemd.uniified-cgroup-hierarchy=no' in F31 after we switch the default. For the loader, we already have a well-established way to set options. I guess it'd make sense to add something like this too sd-boot too, but for me that's an independent issue.

> as you put this in the loader UEFI UUID namespace?

Good point. I'll move it to a different namespace.

> The capitalization of "SystemdOptions" is a bit weird...

Suggestions?

>  KernelCommandLineAppend

As you say, this would then confuse people because it's only about systemd in this patch. I think KernelCommandLineAppend (or Prepend?) could be used for the sd-boot option.

> onshot flavour

OK, will add.

Reply:
Actually, what does `EFI_VENDOR_LOADER` and `EFI_VENDOR_GLOBAL` mean? What namespace would be reasonable for "systemd options"?

Reply:
introduce a new namespace for that. It's just a uuid you can make up.

I am not sure doing a new efi var that feels like a kernel cmdline string but also totally isnt is the best way forward.

I see two options:

1. This could be an option read by the EFI loader stub, and not the EFI boot menu, and is added implicitly to the kernel cmdline.

2. Instead of adding a string formatted option, instead add explicitly settings, i.e. EFI_VENDOR_SYSTEMD:DefaultUnit= for systemd.unit= EFI_VENDOR_SYSTEMD:FileSystemCheck= for systemd.fsck= and so on. i.e. make them high-level variables, and only expose the ones that have the strong usecase. In this case they feel distinct enough not to be confused with a kernel cmdline option.

Reply:
> ntroduce a new namespace for that. It's just a uuid you can make up.

That's what I did.

> This could be an option read by the EFI loader stub, and not the EFI boot menu, and is added implicitly to the kernel cmdline.

This doesn't really solve the goal: I want to have to have a way to amend *systemd* options independently of the boot loader used. And I want this to work "now", i.e. when we are pushing cgroups v2 in Fedora. Some future solution that requires discussion and changes in other projects doesn't really help.

> Instead of adding a string formatted option, instead add explicitly settings, i.e. EFI_VENDOR_SYSTEMD:DefaultUnit= for systemd.unit= EFI_VENDOR_SYSTEMD:FileSystemCheck= for systemd.fsck=

We already expose only specific settings for overriding on the kernel command line, so I think the list that is overridable through EFI should be the same. Otherwise things would be confusing and harder to maintain for no good reason.

OK, I'll try to amend the patch like this.